x/mongo/driver/topology/server.go

```diff
@@ -822,10 +870,100 @@ func isStreamable(srv *Server) bool {
 	return srv.Description().Kind != description.Unknown && srv.Description().TopologyVersion != nil
 }
 
-func (s *Server) check() (description.Server, error) {
+func (s *Server) streamable() bool {
+	return isStreamingEnabled(s) && isStreamable(s)
+}
+
+// getSocketTimeout will return the maximum allowable duration for streaming /
+// polling a hello command during server monitoring.
+func getSocketTimeout(srv *Server) time.Duration {
+	if srv.conn.getCurrentlyStreaming() || srv.streamable() {
+		// If connectTimeoutMS=0, the operation timeout should be infinite.
+		// Otherwise, it is connectTimeoutMS + heartbeatFrequencyMS to account for
+		// the fact that the query will block for heartbeatFrequencyMS
+		// server-side.
+		streamingTO := srv.cfg.connectTimeout
+		if streamingTO != 0 {
+			streamingTO += srv.cfg.heartbeatInterval
+		}
+
+		return streamingTO
+	}
+
+	// The server doesn't support the awaitable protocol. Set the socket timeout
+	// to connectTimeoutMS and execute a regular heartbeat without any
+	// additional parameters.
+	return srv.cfg.connectTimeout
+}
+
+// contextWithSocketTimeout will apply the appropriate socket timeout to the
+// parent context for server monitoring.
+func contextWithSocketTimeout(parent context.Context, srv *Server) (context.Context, context.CancelFunc) {
```

Comment:
Optional: "Socket timeout" seems like a misnomer here (and above in `getSocketTimeout`) because we're applying an operation timeout to the heartbeat, not just a network socket timeout. Consider a different name (e.g. `withHeartbeatTimeout`).